src/category_theory/enriched/basic.lean

```diff
@@ -0,0 +1,435 @@
+/-
+Copyright (c) 2021 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import category_theory.monoidal.types
+import category_theory.monoidal.center
+
+/-!
+# Enriched categories
+
+We set up the basic theory of `V`-enriched categories,
+for `V` an arbitrary monoidal category.
+
+We do not assume here that `V` is a concrete category,
+so there does not need to be a "honest" underlying category!
+
+Use `X ⟶[V] Y` to obtain the `V` object of morphisms from `X` to `Y`.
+
+This file contains the definitions of `V`-enriched categories and
+`V`-functors.
+
+We don't yet define the `V`-object of natural transformations
+between a pair of `V`-functors (this requires limits in `V`),
+but we do provide a presheaf isomorphic to the Yoneda embedding of this object.
+
+We verify that when `V = Type v`, all these notion reduce to the usual ones.
+-/
+
+universes w v u₁ u₂ u₃
+
+noncomputable theory
+
+namespace category_theory
+
+open opposite
+open monoidal_category
+
+variables (V : Type v) [category.{w} V] [monoidal_category V]
+
+/--
+A `V`-category is a category enriched in a monoidal category `V`.
+
+Note that we do not assume that `V` is a concrete category,
+so there may not be an "honest" underlying category at all!
+-/
+class enriched_category (C : Type u₁) :=
+(hom : C → C → V)
+(notation X ` ⟶[] ` Y:10 := hom X Y)
+(id : Π X, 𝟙_ V ⟶ (X ⟶[] X))
+(comp : Π X Y Z, (X ⟶[] Y) ⊗ (Y ⟶[] Z) ⟶ (X ⟶[] Z))
+(id_comp : Π X Y, (λ_ (X ⟶[] Y)).inv ≫ (id X ⊗ 𝟙 _) ≫ comp X X Y = 𝟙 _ . obviously)
+(comp_id : Π X Y, (ρ_ (X ⟶[] Y)).inv ≫ (𝟙 _ ⊗ id Y) ≫ comp X Y Y = 𝟙 _ . obviously)
+(assoc :
+  Π W X Y Z, (α_ _ _ _).inv ≫ (comp W X Y ⊗ 𝟙 _) ≫ comp W Y Z = (𝟙 _ ⊗ comp X Y Z) ≫ comp W X Z
+  . obviously)
+
+notation X ` ⟶[`V`] ` Y:10 := (enriched_category.hom X Y : V)
+
+variables (V) {C : Type u₁} [enriched_category V C]
+
+/--
+The `𝟙_ V`-shaped generalized element giving the identity in a `V`-enriched category.
+-/
+def e_id (X : C) : 𝟙_ V ⟶ (X ⟶[V] X) := enriched_category.id X
+/--
+The composition `V`-morphism for a `V`-enriched category.
+-/
+def e_comp (X Y Z : C) : (X ⟶[V] Y) ⊗ (Y ⟶[V] Z) ⟶ (X ⟶[V] Z) := enriched_category.comp X Y Z
+
+@[simp, reassoc]
+lemma e_id_comp (X Y : C) :
+  (λ_ (X ⟶[V] Y)).inv ≫ (e_id V X ⊗ 𝟙 _) ≫ e_comp V X X Y = 𝟙 (X ⟶[V] Y) :=
+enriched_category.id_comp X Y
+
+@[simp, reassoc]
+lemma e_comp_id (X Y : C) :
+  (ρ_ (X ⟶[V] Y)).inv ≫ (𝟙 _ ⊗ e_id V Y) ≫ e_comp V X Y Y = 𝟙 (X ⟶[V] Y) :=
+enriched_category.comp_id X Y
+
+@[simp, reassoc]
+lemma e_assoc (W X Y Z : C) :
+  (α_ _ _ _).inv ≫ (e_comp V W X Y ⊗ 𝟙 _) ≫ e_comp V W Y Z =
+    (𝟙 _ ⊗ e_comp V X Y Z) ≫ e_comp V W X Z :=
+enriched_category.assoc W X Y Z
```

Comment:
Is it possible to use `restate_axiom` to get these three lemmas?

Reply:
It seems that `restate_axiom` leaves `V` as an implicit argument, which is rather hard to use later. (Perhaps this is why I didn't do it?)

Reply:
Oh I see. Yes, I agree it's better as is.